rfcs/shared/build-system/package-public-api.md

```diff
@@ -0,0 +1,237 @@
+
+**Example:**
+
+```ts
+// @filename package-a/src/index.ts
+
+/**
+ *
+ * @internal
+ */
+export function doSomething(): void {}
+
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/index.d.ts
+
+export declare function hello(): void;
+```
+
+#### `@alpha`
+
+It indicates that an API item is eventually intended to be public, but currently is in an early stage of development
+
+- tooling will trim the declaration from a public API surface.
+- can be exposed only from `/unstable` API surface
+- tolling will NOT TRIM the declaration from `/unstable` API surface.
+
+**Example:**
+
+- unstable api
+
+```ts
+// @filename package-a/src/unstable/index.ts
+
+/**
+ *
+ * @alpha
+ */
+export function doSomething(): void {}
+
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/unstable.d.ts
+export function doSomething(): void;
+export declare function hello(): void;
+```
+
+- stable api
+
+```ts
+// @filename package-a/src/index.ts
+
+// 🚨🚨🚨 NOTE THAT THIS IS PROBABLY A MISTAKE BUT TOOLING WILL COVER YOU
+/**
+ *
+ * @alpha
+ */
+export function doSomething(): void {}
+/**
+ *
+ * @public
+ */
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/index.d.ts
+export declare function hello(): void;
+```
+
+#### `@beta`
+
+It indicates that an API item has been released as a preview or for experimental purposes. Third parties are encouraged to try it and provide feedback. However, a “beta” API should NOT be used in production, because it may be changed or removed in a future version.
+
+- tooling will trim the declaration from a public API surface.
+- can be exposed only from `/unstable` API surface
+- tolling will NOT TRIM the declaration from `/unstable` API surface.
+
+**Example:**
+
+- unstable api
+
+```ts
+// @filename package-a/src/unstable/index.ts
+
+/**
+ *
+ * @beta
+ */
+export function doSomething(): void {}
+
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/unstable.d.ts
+export function doSomething(): void;
+export declare function hello(): void;
+```
+
+- stable api
+
+```ts
+// @filename package-a/src/index.ts
+
+// 🚨🚨🚨 NOTE THAT THIS IS PROBABLY A MISTAKE BUT TOOLING WILL COVER YOU
+/**
+ *
+ * @beta
+ */
+export function doSomething(): void {}
+/**
+ *
+ * @public
+ */
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/index.d.ts
+export declare function hello(): void;
+```
+
+#### `@public`
+
+It indicates that an API item has been officially released, and is now part of the supported contract for a package. Semver versioning is used and enforced thus the API signature cannot be changed without a MAJOR version increment.
+
+- tooling will NOT trim the declaration from a public API surface.
+- CANNOT be exposed from `/unstable` API surface
```

Comment:
>  it seems that everything is implicitly public unless otherwise annotated.?

yeah, API without annotation is implicitly considered as public ->
![image](https://user-images.githubusercontent.com/1223799/174257442-1f93d7f1-b20c-47c9-b54c-8bf6cd8b45d0.png)

We can with enforce it with tooling, so missing annotation will fail on CI/build

![image](https://user-images.githubusercontent.com/1223799/174257607-68b8458b-66a2-4641-812a-6cf15cfd3b6a.png)


>  Do we need this annotation? When would I use it?

It might feel like unnecessary churn for some devs to add this annotation all the time, on the other hand:
- it will force to provide reasonable amount of documentation to public APIs which is what we should be doing anyways
- consistency of using modifiers

personally I have no strong opinions on this, besides having consistency

Reply:
IMO we should enforce it via ESLint to be sure that all APIs are properly annotated.

Reply:
Thanks for the explanation. That really helped.

As you said, it is an extra bit of typing/work but I like the fact that _everything_ would then always be explicit. We uncovered quite a few accidentally public APIs during bug bash so I think it's worth being strict about.